content/post/language_migrations.md

```diff
@@ -0,0 +1,512 @@
+
+1. Our flow matrix contains strictly positive elements, which is a sufficient condition to make it [irreducible](https://en.wikipedia.org/wiki/Irreducibility_(mathematics)) ; from any language there is a way to reach all other ones. Thus, according to [Perron–Frobenius theorem](https://en.wikipedia.org/wiki/Perron%E2%80%93Frobenius_theorem), we are looking for the greatest eigenvalue and the corresponding eigenvector.
+
+2. To find such the dominant eigenvector, we can use the [Power Iteration](https://en.wikipedia.org/wiki/Power_iteration) algorithm. Nevertheless, in addition to be irreducible, the matrix needs to be stochastic and aperiodic.
+
+3. We make our flow matrix become [stochastic](https://en.wikipedia.org/wiki/Stochastic_matrix) by normalizing rows after removing the main diagonal. The greatest eigenvalue is now equal to 1.
+
+4. At last, to make our matrix [aperiodic](https://en.wikipedia.org/wiki/Markov_chain) and well conditioned anyway, there is a famous trick introduced by Larry and Sergey in 1998. It is well explained in [Stanford’s CS246](http://snap.stanford.edu/class/cs246-2013/slides/09-pagerank.pdf) but to make it short, it mainly consists in updating our flow matrix with the following formula:
+
+$$
+P = \\beta P + \\frac{1-\\beta}{N}\\left( \\begin{array}{cccc}
+1 & 1 & ... & 1 \\\\\\
+1 & 1 & ... & 1 \\\\\\
+... & ... & \\ddots & ... \\\\\\
+1 & 1 & ... & 1 \\end{array} \\right)
+$$
+
+where,
+
+* \\(\\beta~\\) is called random walk factor and set to 0.85
+* N   is the number of languages
+
+### Power Iteration
+
+This way, our well conditioned flow matrix contains an approximation of probabilities from switching between languages, and we are now entitled to proceed power iteration. This algorithm simply consists in the following matrix multiplication until convergence to the dominant eigenvector:
+$$
+x_ {i+1} = P\\cdot x_ i
+$$
+
+Below, you will find the code that returns the wanted dominant eigenvector.
+
+```Python
+def power_iteration(A, nb_iterations=100, beta=0.85):
+    u = np.random.rand(len(lang2keep))
+    u = np.reshape(u, (len(lang2keep), 1))
+    A = A * beta + ((1 - beta) / len(lang2keep)) * np.ones((len(lang2keep), len(lang2keep)))
+
+    for _ in range(nb_iterations):
+        u_next = np.dot(A,u)
+        u_next_norm = sum(u_next)
+        u = u_next / u_next_norm
+
+    return u
+
+ power_iteration(transition_matrix)
+```
+<p align="center" class="caption">Power Iteration algorithm, Python.</p>
+
+
+### Most popular languages on GitHub
+
+Finally, here is the reward : the stationary distribution of our [Markov chain](https://en.wikipedia.org/wiki/Markov_chain). This probability distribution is independent from the initial distribution. It gives information about the stability of a random languages switching process. Thus, no matter how popular languages are at the present time, the hypothetical future stationary state stays the same. Here is the popularity ranking of languages used in GitHub:
```

Comment:
~Finally~ At last! Here is the reward

independent of

stability of the process of random ...

on GitHub